File: sdk/typescript/src/bulk.ts
import { HttpClient } from './http.js';
import type {
  CsvImportOptions,
  JsonImportOptions,
  CsvExportOptions,
  JsonExportOptions,
  XlsxExportOptions,
  ImportJobStatus,
  ExportJobStatus,
} from './types.js';

/**
 * Client for bulk import/export operations
 */
export class BulkClient {
  constructor(private readonly http: HttpClient) {}

  // Import Operations

  /**
   * Imports data from a CSV file
   */
  async importCsv(
    tableName: string,
    file: Blob | File,
    options?: CsvImportOptions
  ): Promise<ImportJobStatus> {
    const formData = new FormData();
    formData.append('file', file, 'import.csv');

    if (options?.delimiter) formData.append('delimiter', options.delimiter);
    if (options?.hasHeader !== undefined) formData.append('hasHeader', String(options.hasHeader));
    if (options?.dateFormat) formData.append('dateFormat', options.dateFormat);
    if (options?.trimWhitespace !== undefined)
      formData.append('trimWhitespace', String(options.trimWhitespace));
    if (options?.nullHandling) formData.append('nullHandling', options.nullHandling);
    if (options?.duplicateHandling) formData.append('duplicateHandling', options.duplicateHandling);
    if (options?.keyColumns) formData.append('keyColumns', options.keyColumns.join(','));

    return this.http.postFormData<ImportJobStatus>(
      `/api/bulk/${encodeURIComponent(tableName)}/import/csv`,
      formData
    );
  }

  /**
   * Imports data from a JSON file
   */
  async importJson(
    tableName: string,
    file: Blob | File,
    options?: JsonImportOptions
  ): Promise<ImportJobStatus> {
    const formData = new FormData();
    formData.append('file', file, 'import.json');

    if (options?.jsonPath) formData.append('jsonPath', options.jsonPath);
    if (options?.dateFormat) formData.append('dateFormat', options.dateFormat);
    if (options?.duplicateHandling) formData.append('duplicateHandling', options.duplicateHandling);
    if (options?.keyColumns) formData.append('keyColumns', options.keyColumns.join(','));

    return this.http.postFormData<ImportJobStatus>(
      `/api/bulk/${encodeURIComponent(tableName)}/import/json`,
      formData
    );
  }

  /**
   * Imports data from an NDJSON file
   */
  async importNdjson(
    tableName: string,
    file: Blob | File,
    options?: JsonImportOptions
  ): Promise<ImportJobStatus> {
    const formData = new FormData();
    formData.append('file', file, 'import.ndjson');

    if (options?.dateFormat) formData.append('dateFormat', options.dateFormat);
    if (options?.duplicateHandling) formData.append('duplicateHandling', options.duplicateHandling);
    if (options?.keyColumns) formData.append('keyColumns', options.keyColumns.join(','));

    return this.http.postFormData<ImportJobStatus>(
      `/api/bulk/${encodeURIComponent(tableName)}/import/ndjson`,
      formData
    );
  }

  /**
   * Gets import job status
   */
  async getImportJobStatus(jobId: string): Promise<ImportJobStatus | null> {
    try {
      return await this.http.get<ImportJobStatus>(`/api/bulk/import/${jobId}`);
    } catch (error: unknown) {
      if (error instanceof Error && 'statusCode' in error && (error as { statusCode: number }).statusCode === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Cancels an import job
   */
  async cancelImportJob(jobId: string): Promise<void> {
    await this.http.post(`/api/bulk/import/${jobId}/cancel`);
  }

  // Export Operations

  /**
   * Exports data to CSV format
   */
  async exportCsv(
    tableName: string,
    options?: CsvExportOptions
  ): Promise<ExportJobStatus> {
    return this.http.post<ExportJobStatus>(
      `/api/bulk/${encodeURIComponent(tableName)}/export/csv`,
      options ?? {}
    );
  }

  /**
   * Exports data to JSON format
   */
  async exportJson(
    tableName: string,
    options?: JsonExportOptions
  ): Promise<ExportJobStatus> {
    return this.http.post<ExportJobStatus>(
      `/api/bulk/${encodeURIComponent(tableName)}/export/json`,
      options ?? {}
    );
  }

  /**
   * Exports data to XLSX format
   */
  async exportXlsx(
    tableName: string,
    options?: XlsxExportOptions
  ): Promise<ExportJobStatus> {
    return this.http.post<ExportJobStatus>(
      `/api/bulk/${encodeURIComponent(tableName)}/export/xlsx`,
      options ?? {}
    );
  }

  /**
   * Gets export job status
   */
  async getExportJobStatus(jobId: string): Promise<ExportJobStatus | null> {
    try {
      return await this.http.get<ExportJobStatus>(`/api/bulk/export/${jobId}`);
    } catch (error: unknown) {
      if (error instanceof Error && 'statusCode' in error && (error as { statusCode: number }).statusCode === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Downloads an exported file
   */
  async downloadExport(jobId: string): Promise<Blob> {
    return this.http.getBlob(`/api/bulk/export/${jobId}/download`);
  }

  /**
   * Cancels an export job
   */
  async cancelExportJob(jobId: string): Promise<void> {
    await this.http.post(`/api/bulk/export/${jobId}/cancel`);
  }
}

File: sdk/python/morphdb/bulk.py
"""MorphDB SDK bulk operations client."""

from uuid import UUID

from morphdb.http import HttpClient
from morphdb.models import (
    CsvExportOptions,
    CsvImportOptions,
    ExportJobStatus,
    ImportJobStatus,
    JsonExportOptions,
    JsonImportOptions,
    XlsxExportOptions,
)


class BulkClient:
    """Client for bulk import/export operations."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    # Import operations

    async def import_csv(
        self,
        table_name: str,
        file_content: bytes,
        filename: str = "data.csv",
        options: CsvImportOptions | None = None,
    ) -> ImportJobStatus:
        """Import CSV data into a table."""
        params = {"tableName": table_name}
        if options:
            params.update(options.model_dump(by_alias=True, exclude_none=True))

        data = await self._http.post_file(
            "/api/bulk/import/csv",
            file_content,
            filename,
            "text/csv",
            params,
        )
        return ImportJobStatus.model_validate(data)

    async def import_json(
        self,
        table_name: str,
        file_content: bytes,
        filename: str = "data.json",
        options: JsonImportOptions | None = None,
    ) -> ImportJobStatus:
        """Import JSON data into a table."""
        params = {"tableName": table_name}
        if options:
            params.update(options.model_dump(by_alias=True, exclude_none=True))

        data = await self._http.post_file(
            "/api/bulk/import/json",
            file_content,
            filename,
            "application/json",
            params,
        )
        return ImportJobStatus.model_validate(data)

    async def get_import_status(self, job_id: UUID | str) -> ImportJobStatus:
        """Get import job status."""
        data = await self._http.get(f"/api/bulk/import/{job_id}/status")
        return ImportJobStatus.model_validate(data)

    async def cancel_import(self, job_id: UUID | str) -> ImportJobStatus:
        """Cancel an import job."""
        data = await self._http.post(f"/api/bulk/import/{job_id}/cancel")
        return ImportJobStatus.model_validate(data)

    # Export operations

    async def export_csv(
        self,
        table_name: str,
        options: CsvExportOptions | None = None,
    ) -> ExportJobStatus:
        """Export table data to CSV."""
        body = {"tableName": table_name}
        if options:
            body.update(options.model_dump(by_alias=True, exclude_none=True))

        data = await self._http.post("/api/bulk/export/csv", body)
        return ExportJobStatus.model_validate(data)

    async def export_json(
        self,
        table_name: str,
        options: JsonExportOptions | None = None,
    ) -> ExportJobStatus:
        """Export table data to JSON."""
        body = {"tableName": table_name}
        if options:
            body.update(options.model_dump(by_alias=True, exclude_none=True))

        data = await self._http.post("/api/bulk/export/json", body)
        return ExportJobStatus.model_validate(data)

    async def export_xlsx(
        self,
        table_name: str,
        options: XlsxExportOptions | None = None,
    ) -> ExportJobStatus:
        """Export table data to XLSX."""
        body = {"tableName": table_name}
        if options:
            body.update(options.model_dump(by_alias=True, exclude_none=True))

        data = await self._http.post("/api/bulk/export/xlsx", body)
        return ExportJobStatus.model_validate(data)

    async def get_export_status(self, job_id: UUID | str) -> ExportJobStatus:
        """Get export job status."""
        data = await self._http.get(f"/api/bulk/export/{job_id}/status")
        return ExportJobStatus.model_validate(data)

    async def download_export(self, job_id: UUID | str) -> bytes:
        """Download exported file."""
        return await self._http.get_bytes(f"/api/bulk/export/{job_id}/download")

    async def cancel_export(self, job_id: UUID | str) -> ExportJobStatus:
        """Cancel an export job."""
        data = await self._http.post(f"/api/bulk/export/{job_id}/cancel")
        return ExportJobStatus.model_validate(data)
